### ISSUE TITLE
jacoco hard requires test.exec

### ISSUE THREAD

**Issue by fabienrenaud** [gradle/gradle#5898]
### Summary

`jacocoTestReport` task fails on projects only defining custom test sets (and no "regular" tests in src/test/java). For example, create tests in an integration-test folder (src/integration-test/java) and have no src/test/java folder. The error:
> Unable to read execution data file path/to/build/jacoco/test.exec

### Expected Behavior

Jacoco should have no hard requirements on the existence of test.exec.
Additionally, jacoco should automatically discover any *.exec files in build/jacoco and not bail when test.exec is missing.

### Current Behavior

jacocoTestReport fails processing *.exec files when test.exec is missing
jacocoTestReport needs custom configuration of the executionData attribute when test.exec does not exists. 

### Context

jacocoTestReport completely fails for projects that define integration tests but no (unit) test in src/test.
I reserve test for unit tests. integration tests always go in the integration-test folder by convention.

### Steps to Reproduce (for bugs)

Full and simple repro: https://github.com/fabienrenaud/gradle-jacoco-notestexec

Steps:
1. Create a project, use the org.unbroken-dome.test-sets to configure an integration test directory (e.g: integration-test)
1. Don't write tests in src/test/java. Write tests in src/integration-test/java
1. Run `gradle integrationTest jacocoTestReport`. Notice jacocoTestReport did nothing.
1. Add the following config to gradle:
    ```
    jacocoTestReport {
        sourceSets sourceSets.main
        executionData fileTree(buildDir).include("/jacoco/*.exec")
    }
    ```
1. Run `gradle integrationTest jacocoTestReport`. Notice jacocoTestReport did failing and requiring a test.exec file.

```
~ gradle jacocoTestReport
> Task :jacocoTestReport FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jacocoTestReport'.
> Unable to read execution data file /Users/frenaud/projects/me/gradle-jacoco-notestexec/build/jacoco/test.exec
```

### Your Environment

```
------------------------------------------------------------
Gradle 4.8.1
------------------------------------------------------------

Build time:   2018-06-21 07:53:06 UTC
Revision:     0abdea078047b12df42e7750ccba34d69b516a22

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_131 (Oracle Corporation 25.131-b11)
OS:           Mac OS X 10.12.6 x86_64
```

**Comment by fabienrenaud** [comment]
Even if it wires to the default test task:
 * why failing jacocoTestReport if test.exec is missing? isn't there a state from the test task jacoco could read to check there are no tests for the test task?
 * why not wiring automatically to all tasks of Test type?
 * why wiring to any task of Test type at all? why not just reading all *.exec files in build/jacoco/ ?

**Comment by liorhar** [comment]
I believe I run into the same issue. I'm trying to split the DB tests from the main test task. I added a `dbTest` task and jacoco generates dbTest.exec file successfully. I configured the `executionData` property to include `*.exec` files. However, when running `gradle dbTest jacocoTestReport` it will fail on the missing `test.exec` file.
Is there a way to generate the jacoco report for an alternate test task without running the main test task?

**Comment by arvindn05** [comment]
Doings the following worked for me. Note the '=' like mentioned above.
Basically i include both test and integration test data by default and while generating the report i filter the execution data to only pick coverage files which exist.
```
subprojects {
    jacocoTestReport {
        executionData(test, integrationTest)
    }
}

def allTestCoverageFile = "$buildDir/jacoco/allTestCoverage.exec"
task jacocoMergeTestData(type: JacocoMerge) {
    destinationFile = file("$buildDir/jacoco/testCombo.exec")
    executionData = files(subprojects.jacocoTestReport.executionData).filter { f -> f.exists() }
}
task jacocoRootReport(type: JacocoReport) {
    dependsOn jacocoMergeTestData
    description = 'Generates an aggregate report from all subprojects'

    additionalSourceDirs = files(subprojects.sourceSets.main.allSource.srcDirs)
    sourceDirectories = files(subprojects.sourceSets.main.allSource.srcDirs)
    classDirectories = files(subprojects.sourceSets.main.output)
    executionData = files(subprojects.jacocoTestReport.executionData).filter { f -> f.exists() }
    
    reports {
        xml.enabled = false
        html.enabled = true
    }
}
```
Hope this helps :)

**Comment by ghilainm** [comment]
This does not work anymore in Gradle 6.0. This is forbidden to override executionData. 

> The JacocoReportBase.setExecutionData(FileCollection) method has been deprecated. This is scheduled to be removed in Gradle 6.0. Use getExecutionData().from(...)

What is the recommended way of overriding execution data then?

**Comment by RothAndrew** [comment]
@ghilainm 

> What is the recommended way of overriding execution data then?

```
jacocoTestReport {
    getExecutionData().setFrom(fileTree(buildDir).include("/jacoco/*.exec"))
}
```

**Comment by wujek-srujek** [comment]
I have recently been hit by this issue, and dug deeper (the workaround didn't work for me immediately as our project structure is a bit more complex than the example) and found out that the Test Sets plugin creates a JacocoReport task for each new source set. Each such new task uses its own *.exec file as input from the corresponding custom Test tasks, and generates its own report. (the reports use inconsistent destination file/folder naming, see https://github.com/unbroken-dome/gradle-testsets-plugin/issues/100).
This knowledge helped me understand what is going on and implement a solution working for our project.

**Comment by yuvalprtn** [comment]
@wujek-srujek  can you share the solution?

**Comment by wujek-srujek** [comment]
Yes. Our project is a multimodule project. Here are the most relevant parts, in Kotlin DLS:

gradle.build.kts:

    apply(plugin = "jacoco")

    val aggregateJacocoTestReport = tasks.create("aggregateJacocoTestReport", JacocoReport::class) {
        executionData.setFrom(fileTree(projectDir).include("**/*.exec"))
        // will further be configured by subprojects loop
    }

    tasks.create("build") {
        dependsOn(aggregateJacocoTestReport)
    }

    subprojects notLeaf@{
        // apply the configuration to leaf modules only, i.e. not 'grouping' folders
        if (subprojects.size > 0) {
            return@notLeaf
        }

        apply(plugin = "jacoco")
        apply(plugin = "org.unbroken-dome.test-sets")

        configure<TestSetContainer> {
            "integrationTest"()
        }

        val check by tasks
        val build by tasks

        tasks.withType<Test> {
            useJUnitPlatform()
            check.dependsOn(this)
        }

        // all test tasks have their own JaCoCo report, so let's create one that aggregates them
        val jacocoAllTestReport = tasks.create("jacocoAllTestReport", JacocoReport::class) {
            executionData.setFrom(fileTree(buildDir).include("**/*.exec"))
        }

        // add current project's production source set to the custom JaCoCo report tasks
        // jacocoAllTestReport gets the source set from the project it's in
        // aggregateJacocoTestReport gets all production source sets
        project.configure<SourceSetContainer> {
            val mainSourceSet = getByName(SourceSet.MAIN_SOURCE_SET_NAME)
            jacocoAllTestReport.sourceSets(mainSourceSet)
            aggregateJacocoTestReport.sourceSets(mainSourceSet)
        }

        jacocoAllTestReport.dependsOn(check)
        build.dependsOn(jacocoAllTestReport)

        aggregateJacocoTestReport.dependsOn(check)
    }

To sum up:
* We have a global custom JaCoCo report for the whole project and the custom 'build' task depends on it. This causes calls to './gradlew build' also execute the root 'build' task, which in turn executes the global report task.
* Every submodule gets a custom 'jacocoAllTestReport'. It is made dependent on the 'check' task, which in turn is dependent on all 'test' tasks (like 'integrationTest' in our case).
* The global report task gets the main source set from each module, and is made dependent on each submodule's 'check'.
* When we want coverage report for a module, we call its 'jacocoAllTestReport' task. We can still the other tasks if we need a report for each test source set separately, be we normally don't need it.

**Comment by ashwini-desai** [comment]
Our's is also a multi-module gradle project with integration tests in one of them. We ran into the same problem of not being able to merge ```**/jacoco/test.exec``` files. 

Solution below worked for us:
```
subprojects {
    apply(plugin: 'org.jetbrains.kotlin.jvm')

    repositories {
        jcenter()
        mavenCentral()
   }
}

task codeCoverageReport(type: JacocoReport) {

    // Gather execution data from all subprojects
    executionData fileTree(project.rootDir.absolutePath).include("**/build/jacoco/*.exec")

    // Add all relevant sourcesets from the subprojects
    subprojects.each {
        sourceSets it.sourceSets.main
    }

    reports {
        xml.enabled true
        html.enabled true
        csv.enabled false
    }
}

// always run the tests before generating the report
codeCoverageReport.dependsOn {
    subprojects*.test
}
```

Run ```gradle codeCoverageReport```.  We tried making available all source-sets to the integration test module in many ways but nothing other than the above solution worked. To make available sourcesets of all modules, looping over subprojects and accumulating sourcesets worked. **subprojects.sourceSets.main.allSource.srcDirs** did not work.

**Comment by ddsultan** [comment]
> The JaCoCo plugin creates the `jacocoTestReport` task and wires it to the default `test` task. I know it's not obvious, but calling [`executionData(Object...)`](https://docs.gradle.org/current/javadoc/org/gradle/testing/jacoco/tasks/JacocoReportBase.html#executionData-java.lang.Object...-) adds files to the default `FileCollection` which already contains `test.exec`.
> 
> Calling `setExectionData(FileCollection)` avoids this issue (note the `=`):
> 
> ```groovy-gradle
> jacocoTestReport {
>     sourceSets sourceSets.main
>     executionData = fileTree(buildDir).include("/jacoco/*.exec")
> }
> ```

I am getting the following error with `gradle 6.1` and `jacoco  0.8.5`:

```bash
Cannot set the value of read-only property 'executionData' for task ':jacocoTestReport'
```

**Comment by JMartinetti20** [comment]
> > The JaCoCo plugin creates the `jacocoTestReport` task and wires it to the default `test` task. I know it's not obvious, but calling [`executionData(Object...)`](https://docs.gradle.org/current/javadoc/org/gradle/testing/jacoco/tasks/JacocoReportBase.html#executionData-java.lang.Object...-) adds files to the default `FileCollection` which already contains `test.exec`.
> > Calling `setExectionData(FileCollection)` avoids this issue (note the `=`):
> > ```
> > jacocoTestReport {
> >     sourceSets sourceSets.main
> >     executionData = fileTree(buildDir).include("/jacoco/*.exec")
> > }
> > ```
> 
> I am getting the following error with `gradle 6.1` and `jacoco 0.8.5`:
> 
> ```shell
> Cannot set the value of read-only property 'executionData' for task ':jacocoTestReport'
> ```

This was answered in the thread:



> @ghilainm
> 
> > What is the recommended way of overriding execution data then?
> 
> ```
> jacocoTestReport {
>     getExecutionData().setFrom(fileTree(buildDir).include("/jacoco/*.exec"))
> }
> ```

This is working for me now and my multi-test gradle project.

**Comment by hpatel123** [comment]
@ashwini-desai I would request you to provide whole code base for jacoco as I'm trying to fix but nothing works for me.

**Comment by pa1maurya** [comment]
@hpatel123 is your error fixed ? Its not solved for me as well
